### ISSUE TITLE
Viewport offset with spheric mirror distortion 

### ISSUE THREAD

**Issue by Jimi-Harrold** [Stellarium/stellarium#3499]
**Is your feature request related to a problem? Please describe.**
We use Stellarium with an inflatable planetarium as part of our universities outreach program. In general we have had great success using Stellarium with the spheric mirror distortion and the scripting language. However, by default Stellarium centers selected objects to the middle of the screen, which ends up being the top of the planetarium. This causes may strained necks in the children we present to. Ideally we would like centred objects to be placed in the lower quarter of the screen when centred. 

We have tried commands (both before and after a command such as `core.moveToSelectedObject()` ) such as: `StelMovementMgr.setViewportVerticalOffsetTarget()`, `StelMovementMgr.moveViewport()`, `StelMovementMgr.panView()`, `StelMovementMgr.smoothPan()`, `StelMovementMgr.moveToAltAzi()`, and `core.moveToAltAzi()`. All these commands work in a vacuum, but when combined with moving to different stars/constellations often result in either us having to move the constellation overhead and then down to where we want it (e.g. `setViewportVerticalOffsetTarget`, `moveViewport`, `smoothPan`) or end up with some minor "jolts" when movement starts (e.g. using `moveToAltAzi` with an offset applied to the objects altitude) both of which are not ideal in a planetarium with easily nauseated children! 

Ideally we would just want to apply a permanent offset to the viewport so running a single command like `core.moveToSelectedObject()` would place the centred object lower down, I was hoping a command such as `setViewportVerticalOffsetTarget` would do this but the value appears to be reset when a new object is centred. We've also tried various config adjustments such as using "`viewportCenterY`" or "`viewport_center_offset_y`" but we haven't been able to sea any changes take place as a result of altering these values. 

**Describe the solution you'd like**
Either a command or a config entry that adjusts where a centred object appears on screen, so that running a command like "`core.moveToSelectedObject()`" centres the object at some point other than the exact middle of the viewport.

**Describe alternatives you've considered**

1. Setting a viewport offset both before and after the  `core.moveToSelectedObject()` command, which always seems to be reset by the `core.moveToSelectedObject()` command.
2. Using `StelMovementMgr.moveToAltAzi()` or `core.moveToAltAzi()` with a set adjustment to the objects altitude, which often results in a jarring jump at the start of a new move (this is perhaps fixable but we've not found a way).
3. Using config entries such as "`viewportCenterY`" and "`viewport_center_offset_y`" in an attempt to offset the centre of the viewport, neither command seemed to have any noticeable effect.

**Additional context**
We use spheric projection with the config values:
```
[spheric_mirror]
distorter_max_fov                         = 180
dome_radius                               = 5.0
flag_use_ext_framebuffer_object           = false
flip_horz                                 = false
flip_vert                                 = false
image_distance_div_height                 = 2.67
mirror_position_x                         = 0
mirror_position_y                         = 5
mirror_position_z                         = 0
mirror_radius                             = 0.37
projector_alpha                           = 0
projector_delta                           = 10.6
projector_gamma                           = 0.3
projector_phi                             = 0
projector_position_x                      = 0
projector_position_y                      = 4
projector_position_z                      = 0
texture_triangle_base_length              = 8
```

**Comment by gzotti** [comment]
I have been surprised by the "center in viewport" myself in a recent show in a real planetarium. Indeed, the default behaviour of select&move to center is not ideal in a planetarium but targets the 99.99678% of other users with regular screens :-P 

In 2016 I have added a scriptable command `StelMovementMgr.setInhibitAllAutomoves(true);`
which is useful in a context where centering is a no-go: Planetarium, static panorama wall, ... You may want to add this to your startup.ssc (which is run at start).

This only highlights the object. In a planetarium this is probably the best action.

**Comment by Jimi-Harrold** [comment]
> I have been surprised by the "center in viewport" myself in a recent show in a real planetarium. Indeed, the default behaviour of select&move to center is not ideal in a planetarium but targets the 99.99678% of other users with regular screens :-P
> 
> In 2016 I have added a scriptable command `StelMovementMgr.setInhibitAllAutomoves(true);` which is useful in a context where centering is a no-go: Planetarium, static panorama wall, ... You may want to add this to your startup.ssc (which is run at start).
> 
> This only highlights the object. In a planetarium this is probably the best action.

Hi gzotti,

Thank you for your swift response. I forgot to add that we tried that command as well but due to the nature of our shows (which are delivered to primary schools often with high amounts of special educational needs) we require the children to all sit facing one direction, it is very difficult for wheelchairs and such to pivot which is partly why we needed the centres object lower in the first place! 

I do understand this is a small issue affecting a small group of people. If the config options or viewport adjustment should have worked as I described (I was really struggling to find documentation on a lot of the spheric mirror and viewport options) but aren't,  I am happy to try tweaking it and opening a pull request. This may be easiest because as you said, we are a very small portion of the overall userbase. 

Thanks again for getting back so quickly!

**Comment by gzotti** [comment]
I have never used or configured such a spherical mirror setup, so I cannot comment whether viewport changes work here. The StelMovementMgr is a class asking for a thorough rewrite, here at least I won't attach yet another option in the near future, sorry. 
You can still move the sky view with keyboard or mouse after selection. There is just no automove/autozoom.

**Comment by Jimi-Harrold** [comment]
Unfortunatly due to the fact we use undergraduate volunteers who may not be too familiar with Stellarium, manual controls are not ideal for us. Thankfully I think I have something that works well enough for our dome, I'll document it bellow and close the issue. 

Looking around the source code, I have managed to implement a workaround:
```
StelMovementMgr.setViewportVerticalOffsetTarget (-30)
function gotoObject(object_name, use_pointer, duration){
	var pos = core.getObjectInfo(object_name);
	core.moveToRaDecJ2000(pos["ra"],pos["dec"],duration)
	core.selectObjectByName(object_name,use_pointer)
}
```

With some example uses:

```
gotoObject("Mizar",true,1)
core.wait(3)
gotoObject("M42",true,1)
core.wait(3)
gotoObject("Aldebaran",true,1)
```

As far as I can tell this works because in `StelMovementMgr` when the automove updates, if an object is selected to move to it takes into account the viewpoint offset to force the object back into the center of the screen, but skips this check if  raw coordinates were given (only the y axis offset, line 1122 in `StelMovementMgr` in the function `StelMovementMgr::updateVisionVector`  for reference). As we don't need to manually add an offset now, we can use the command `setViewportVerticalOffsetTarget`, allowing us to use `moveToRaDecJ2000` directly which skips the coordinate transform and the nauseating jump every time a new move is issued. Selection of the object needs to happen after as the move command nullifies the selected object as far as I can tell.

Also worth noting, I believe the config command `viewport_center_offset_y` should work with this too, the problem that occurred for us that meant it was hard to tell if it was working is that when you center an object it overrides the offset, meaning it always seemed as if the parameter does nothing on a brief test.

**Comment by gzotti** [comment]
Good you found a solution. I just found we did already something to remedy the "zenith centering". In the RemoteControl panel, when selecting, below the search field there are settings what to do. "Just mark the object" is the least surprising option, useful in a planetarium. Then move view with usual methods if required. (In some planetaria there is a "freeze azimuth orientation" rule to avoid directional confusion. This may not be your situation.)
